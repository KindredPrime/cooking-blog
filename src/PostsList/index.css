.PostsList {
  margin-top: 50px;
}

.PostsList__blog-post {
  margin-bottom: 50px;
}

.PostsList__blog-post * {
  margin: 0 0 5px 0;
}

.PostsList__title {
  font-weight: bold;
  font-size: 20px;
}

.PostsList__title:focus, .PostsList__title:hover {
  text-decoration: underline;
}