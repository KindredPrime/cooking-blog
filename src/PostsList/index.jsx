import { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { sortEntities, formatDate, isOnLastPage } from '../util';
import TransitionButtons from '../TransitionButtons/index';
import './index.css';

class PostsList extends Component {
  state = {
    page: 1,
    sortedPosts: [],
    displayButtons: false
  };

  static defaultProps = {
    pageLimit: 10,
    initialBlogPosts: []
  };

  componentDidMount() {
    const { initialBlogPosts, pageLimit } = this.props;

    this.setState({
      sortedPosts: sortEntities(initialBlogPosts),
      displayButtons: initialBlogPosts.length > pageLimit
    });
  }

  render() {
    const { pageLimit } = this.props;
    const { page, sortedPosts, displayButtons } = this.state;

    const postsToRender = sortedPosts.slice((page-1) * pageLimit, page * pageLimit);

    return (
      <section className="PostsList">
        <header>
          <h2>Posts</h2>
        </header>

        <ul>
          {postsToRender.map((post) => {
            const { id, title, lastEdited } = post;

            return (
              <li key={`post-${id}`} className="PostsList__blog-post">
                <Link className="PostsList__title" to={`/blog-posts/${id}`}>{title}</Link>
                <p className="timestamp">Last edited: {formatDate(lastEdited)}</p>
              </li>
            );
          })}
        </ul>

        {displayButtons
          && <TransitionButtons
            isAtTheBeginning={page === 1}
            isAtTheEnd={isOnLastPage(sortedPosts, page, pageLimit)}
            handlePrevClick={() => {
              this.setState({
                page: page - 1
              });
            }}
            handleNextClick={() => {
              this.setState({
                page: page + 1
              });
            }}
          />}
      </section>
    );
  }
}

PostsList.propTypes = {
  initialBlogPosts: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      lastEdited: PropTypes.instanceOf(Date).isRequired,
      title: PropTypes.string.isRequired,
      content: PropTypes.string.isRequired,
      author: PropTypes.shape({
        id: PropTypes.number.isRequired,
        username: PropTypes.string.isRequired
      }).isRequired
    })
  ),
  pageLimit: PropTypes.number
};

export default PostsList;