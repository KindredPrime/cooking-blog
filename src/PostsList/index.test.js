import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom';
import PostsList from './index';

describe('PostsList Component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <BrowserRouter>
        <PostsList />
      </BrowserRouter>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });
});