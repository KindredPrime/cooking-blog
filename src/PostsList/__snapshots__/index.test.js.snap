// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsList Component renders the UI as expected 1`] = `
<body>
  <div>
    <section
      class="PostsList"
    >
      <header>
        <h2>
          Posts
        </h2>
      </header>
      <ul>
        <li
          class="PostsList__blog-post"
        >
          <a
            class="PostsList__title"
            href="/blog-posts/8"
          >
            Title8
          </a>
          <p
            class="timestamp"
          >
            Last edited: 
            Jan 06 2021 12:22:39
          </p>
        </li>
        <li
          class="PostsList__blog-post"
        >
          <a
            class="PostsList__title"
            href="/blog-posts/6"
          >
            Title6
          </a>
          <p
            class="timestamp"
          >
            Last edited: 
            Jan 05 2021 14:22:39
          </p>
        </li>
        <li
          class="PostsList__blog-post"
        >
          <a
            class="PostsList__title"
            href="/blog-posts/4"
          >
            Cooking Basics for Beginners
          </a>
          <p
            class="timestamp"
          >
            Last edited: 
            Jan 05 2021 08:28:28
          </p>
        </li>
      </ul>
      <div
        class="TransitionButtons"
      >
        <button
          class="TransitionButtons__prev"
          disabled=""
          type="button"
        >
          Previous
        </button>
        <button
          class="TransitionButtons__next"
          type="button"
        >
          Next
        </button>
      </div>
    </section>
  </div>
</body>
`;

exports[`PostsList Component switches to the next group of blog posts after clicking 'Next' 1`] = `
<body>
  <div>
    <section
      class="PostsList"
    >
      <header>
        <h2>
          Posts
        </h2>
      </header>
      <ul>
        <li
          class="PostsList__blog-post"
        >
          <a
            class="PostsList__title"
            href="/blog-posts/3"
          >
            5 Easy Cooking Tips
          </a>
          <p
            class="timestamp"
          >
            Last edited: 
            Dec 01 2020 12:54:31
          </p>
        </li>
      </ul>
      <div
        class="TransitionButtons"
      >
        <button
          class="TransitionButtons__prev"
          type="button"
        >
          Previous
        </button>
        <button
          class="TransitionButtons__next"
          disabled=""
          type="button"
        >
          Next
        </button>
      </div>
    </section>
  </div>
</body>
`;
